### ISSUE TITLE
Named goals in {…}

### ISSUE THREAD

**Issue by nomeata** [rocq-prover/rocq#8026]
Now that we have the `2: {…}` syntax, I was hoping I can do this:

```
Lemma foo (n: nat) (P : nat -> Prop):
  P n.
Proof.
  intros.
  refine (nat_ind _ ?[Base] ?[Step] _).
  [Base]: { … }.
  [Step]: { … }.
Qed.
```
but it is not yet supported, it seems. Is there a plan to support this?
EDIT: supported since 8.9

And, as a next step, is there a way of indicating the name for a named goals already in the lemma, so that this would work (note that the assumptions of `nat_ind_named` have names):
```
Lemma nat_ind_named:
  forall (P : nat -> Prop) (n: nat)
  (base : P 0)
  (step : forall n, P n -> P (S n)),
  P n.
Proof. intros. apply nat_ind; assumption. Qed.
  
Lemma foo (n: nat) (P : nat -> Prop):
  P n.
Proof.
  intros.
  apply nat_ind_named.
  [base]: { … }.
  [step]: { … }.
Qed.
```

**Comment by Zimmi48** [comment]
@nomeata Thanks for your request. This had been my plan following the introduction of `2: { … }` but I was waiting for user requests to confirm the need. See #8112 for the first part of your request.

For the second part, I would indeed support such a feature (it would be very useful to have this at least for auto-generated induction principles for inductive types), but I'm afraid there are deeper unanswered design questions about the lives of names in general. See also #7997 and #268. I'm sure that @ppedrot can comment on the present proposal as well.

While waiting for a big redesign that would help solving all the names issues, what about a tactic that would allow renaming sub-goals, for instance:

```
Lemma foo (n: nat) (P : nat -> Prop):
  P n.
Proof.
  intros.
  apply nat_ind_named; name_goals [ base | step ].
```

where `name_goals [ base | step ]` would be equivalent to `[ refine ?[base] | refine ?[step] ]`?

I'm not sure it would represent such a big improvement over just using a plain `refine` though...

**Comment by nomeata** [comment]
I guess one would wrap it in a tactic in either case, so the gain is small - but but completely zero, as a `named_goal` tactic in the documentation would highlight this pattern!

**Comment by Zimmi48** [comment]
#8112 has been merged, and the [documentation for this feature](https://gitlab.com/coq/coq/-/jobs/artifacts/master/file/_install_ci/share/doc/coq/sphinx/html/proof-engine/proof-handling.html?job=documentation#navigation-in-the-proof-tree) contains an example with `eexists` and a note that a goal can be renamed by using `refine ?[ident]`.

I am not sure of the next steps: I'd rather have users use Tactic Notation to define stuff like "name_goals" but unfortunately this is not possible as of today because of #8344 and #8342. If these two issues were solved, one could write something like `Tactic Notation "goals" "[" ident(x1) "|" ident(x2) "]" := [> refine ?[x1] | refine ?[x2] ]` and probably even go further with recursive notations.

As for the automatic naming of goals, this must be done while keeping in mind the global picture of name generation and preservation. I would like to get comments from @ppedrot and @herbelin on this question.

**Comment by jashug** [comment]
TL;DR: As long as it works with `-mangle-names` and there is a way to give names explicitly, auto-naming goals shouldn't make the current situation worse, and it is very much consistent with current Ltac style as far as I understand it.

Assuming the names of goals/evars must be unique they will have to be refreshed, so the names might end up with numeric suffixes unexpectedly. This already happens in this example with the two goals/evars being named `?Goal` and `?Goal0` as far as I can tell, though these names might not be accessible from Ltac.

This automatic naming and renaming, being based on several arbitrary and heuristic choices, makes proof scripts hard to read without stepping through the proof and makes proofs fragile with respect to future changes in the naming algorithm. It is also entirely consistent with the current treatment of variables in context in Ltac, and many users don't need to make their proofs maintainable across Coq versions or easy for a human to read.

As long as the assignment of names to goals goes through `Namegen.next_name_away` and related functions, `-mangle-names` provides a centralized way to ensure that proofs aren't depending on those automatically generated names while keeping most of the benefits given by smart automatic naming including simplicity while prototyping and ease of tracking goals while writing proofs.
(With the caveat that `-mangle-names` is not really ready for use yet, due to other ways Coq is depending on automatic naming in unexpected places.)

The alternate model where hypotheses have a single name and using apply introduces a goal of that name, or fails if the name is already used, doesn't seem to have much support.

**Comment by Zimmi48** [comment]
In fact, the "block destrcturing" intro-pattern that I was mentioning above was also inspiring me a prefix/suffix-based solution, e.g. `apply lemma naming goals with suffix _foo`.

**Comment by cpitclaudel** [comment]
With https://github.com/rocq-prover/rocq/pull/20809 and related PRs being merged, I believe this is now fixed :)  Can I go ahead and close the issue?
